spec/models/csv_file_upload_spec.rb

```diff
@@ -0,0 +1,43 @@
+require "rails_helper"
+
+RSpec.describe CsvFileUpload do
+  let(:csv) do
+    <<~CSV
+      foo | bar | baz
+      1   | 2   | 3
+    CSV
+  end
+  let(:file) { double("File") }
+  let(:params) do
+    {
+      csv: file,
+    }
+  end
+
+  let(:upload) { described_class.new(params, :csv) }
+  let(:result) { upload.to_csv }
+
+  before do
+    allow(file).to receive(:read) { csv.gsub(/ *\| */, ",") }
+  end
+
+  it "returns a parsed csv" do
+    expect(result).to be_a(CSV::Table)
+    expect(result.first.to_h).to eq({"foo" => "1", "bar" => "2", "baz" => "3"})
+  end
+
+  context "when the file has a BOM prefix" do
+    let(:bom) { "\uFEFF".force_encoding("binary") }
+    let(:csv) do
+      bom + <<~CSV
```

Comment:
There's an encoding bug here caused by concatenating an `ASCII_8BIT` string with a `UTF_8` one. It happens to work because the CSV data does not contain any non-ASCII codepoints, but if you change `3` to `hëllo`, for example, you'll see the error.

This is fixed by removing `.force_encoding("binary")`, but it exposes the bug I mentioned above, with a confusing failure message:

```
       expected: {"bar"=>"2", "baz"=>"hëllo", "foo"=>"1"}
            got: {"bar"=>"2", "baz"=>"hëllo", "foo"=>"1"}

       (compared using ==)

       Diff:
       @@ -1,4 +1,4 @@
        "bar" => "2",
        "baz" => "hëllo",
       -"foo" => "1",
       +"foo" => "1",
```

(This failure happens if you don't modify `csv` too.) This is because, with the `force_encoding` call present in the definition of `bom`, `csv` is an `ASCII_8BIT` string. If you remove `force_encoding`, then `csv` becomes a `UTF_8` string, and `CsvFileUpload` fails to remove the BOM from it, so the first key in the resulting hash is actually `"<U+FEFF>foo"`, which you can't see because U+FEFF is a non-printing character.

Implementing `CsvFileUpload#body` as above fixes this bug, so I'd recommend including some non-ASCII data in the tests, and making `file.read` return both `ASCII_8BIT` and `UTF_8` strings to make sure the bug is fixed.

Reply:
That's really clear and useful, ta 👍